internal/pkg/cli/deploy/env.go

```diff
@@ -100,11 +141,115 @@ func NewEnvDeployer(in *NewEnvDeployerInput) (*envDeployer, error) {
+		// However, when clients enable CloudFront, they would need to update all their existing records to now point to the distribution.
+		// Hence, we force users to use `alias` and let the records be written in the environment stack instead.
+		if err := d.envDescriber.ValidateCFServiceDomainAliases(); err != nil {
+			return err
+		}
+	}
+
+	if mft.CDNEnabled() && aws.BoolValue(mft.CDNConfig.Config.TerminateTLS) {
+		// ensure all services _are not_ doing http->https redirect
+		if err := d.verifyALBWorkloadsDontRedirect(ctx); err != nil {
+			return fmt.Errorf("can't enable TLS termination on CDN: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// verifyALBWorkloadsDontRedirect verifies that none of the ALB Workloads
+// in this environment have a redirect in their HTTPWithDomain listener.
+// If any services redirect, an error is returned.
+func (d *envDeployer) verifyALBWorkloadsDontRedirect(ctx context.Context) error {
+	params, err := d.envDescriber.Params()
+	if err != nil {
+		return fmt.Errorf("get env params: %w", err)
+	}
+
+	services := strings.Split(params[cfnstack.EnvParamALBWorkloadsKey], ",")
+	g, ctx := errgroup.WithContext(ctx)
+
+	var badServices []string
+	var badServicesMu sync.Mutex
+
+	for i := range services {
+		svc := services[i]
+		g.Go(func() error {
+			redirects, err := d.lbServiceRedirects(ctx, svc)
+			switch {
+			case err != nil:
+				return fmt.Errorf("verify service %q: %w", svc, err)
+			case redirects:
+				badServicesMu.Lock()
+				defer badServicesMu.Unlock()
+				badServices = append(badServices, svc)
+			}
+
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+	if len(badServices) > 0 {
+		sort.Strings(badServices)
+		return fmt.Errorf("HTTP traffic redirects to HTTPS in %v %v.\nSet http.redirect_to_https to false for %v and redeploy %v.",
+			english.PluralWord(len(badServices), "service", "services"),
+			english.OxfordWordSeries(badServices, "and"),
+			english.PluralWord(len(badServices), "that service", "those services"),
+			english.PluralWord(len(badServices), "it", "them"),
+		)
+	}
+
+	return nil
+}
+
+// lbServiceRedirects returns true if svc's HTTP listener rule redirects.
+func (d *envDeployer) lbServiceRedirects(ctx context.Context, svc string) (bool, error) {
+	stackDescriber := d.newServiceStackDescriber(svc, d.envManagerSession)
+	resources, err := stackDescriber.Resources()
+	if err != nil {
+		return false, fmt.Errorf("get stack resources: %w", err)
+	}
+
+	ruleARN := ""
+	for _, res := range resources {
+		if res.LogicalID == svcStackResourceHTTPListenerRuleLogicalID {
+			ruleARN = res.PhysicalID
+		}
+	}
+	if ruleARN == "" {
+		return false, fmt.Errorf("resource %q not present", svcStackResourceHTTPListenerRuleLogicalID)
+	}
+
+	lbDescriber := d.newLBDescriber(d.envManagerSession)
+	isRedirect, err := lbDescriber.ListenerRuleIsRedirect(ctx, ruleARN)
```

Comment:
qq:

If we go with the route of fetching the physical resource ID from the service stack, then is it safe for us to make the assumption that `HTTPListenerRuleWithDomain` must have the `redirect` action?

Reply:
I think so - here's my reasoning:

We enable HTTPS for LBWS if a [domain or public certs have been imported](https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/deploy/cloudformation/stack/lb_web_svc.go#L72-L88). If HTTPS is enabled, then the service stack [uses](https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/template/templates/workloads/partials/cf/alb.yml#L82-L86) the https-listener, which _always_ creates [`HTTPListenerRuleWithDomain`](https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/template/templates/workloads/partials/cf/https-listener.yml#L74). Then, based on -----------------

great question - thanks for making me think about it :) i think we should modify the check [here](https://github.com/aws/copilot-cli/pull/4017/files#diff-924fba6d7ac2c8e2b099ec2776e78f958c7b225d5484001289b0e62283df3db1R169) to ensure we only run this check when a domain or public cert is imported to the env. once i change that, does this logic make sense to you? we can be safe assuming that `HTTPListenerRuleWithDomain` will always be there because we know the environment is using HTTPS.